torch/nn/utils/stateless.py

```diff
@@ -37,21 +36,83 @@ def _setattr(self, name: str, value: Any) -> None:
+        {'tied_foo': 'foo'}.
+
+    ex: If there aren't any tied weights and the user passed values for every parameter and buffer, this will return a
+        map where every name maps to an empty set: {'l1.weight': set(), 'l1.bias': set(), ...}
 
+    ex: The map only contains values that a user is reparamaterizing. For example, if module = nn.Linear(...) and the
+        user only passed a new value for 'bias', this looks returns: {'bias': set()}
 
-def _create_swap_params(params_and_buffers, replaced_tensors_map):
-    def _swap_parameters(module, tensor_name: str, full_path: str, tensor: Tensor) -> None:
+    This is useful because we will start by reparamaterizing all the keys of params_and_buffers, then all the key from
+    this returned dictionary.
+    """
+
+    # The basic algorithm looks like:
+    #   - index all weights by their original tensor value to find tied weights
+    #     - when we encounter a weight not used by the user, we save it in a set (second element in the tuple)
+    #     - when we run into a weight used by the user, we save that separate from the set as the first element in the tuple
+    #     - ending map looks like {tensor: (name_given_by_user, set(all_tied_names)}
+    #   - then loop through the values of this map (name_given_by_user and set(all_tied_names))
+    #     - for each element of all_tied_names, add {tied_name: name_given_by_user} to a new map
+
+    names = params_and_buffers.keys()
+    weight_to_name_and_tied_names: Dict[torch.Tensor, Tuple[Optional[str], Set[str]]] = {}
+
+    # create a map keyed by tensor value so that tied weights get mapped to the same key. The value is the interesting
+    # part at the end it's (used_name, (tied_names)).
+    # For example, in the first example where there's tied weights self.foo and self.tied_foo and the user passes a
+    # value for self.foo, this will return {torch.Tensor(...): ('foo', set('tied_foo'))}
+    def add_to_name_map(n: str, t: torch.Tensor):
+        # if the tensor hasn't been seen before, add it to the map
+        if t not in weight_to_name_and_tied_names:
+            weight_to_name_and_tied_names[t] = (n, set()) if n in names else (None, {n})
+            return
+
+        # if the name is not used by the user, we add it to the tied set
+        if n not in names:
+            weight_to_name_and_tied_names[t][1].add(n)
+            return
+
+        # check that the user didn't pass two different tensors for the same tied weight
+        first_seen_name = weight_to_name_and_tied_names[t][0]
+
+        # if they didn't pass multiple names for tied weights or used the same tensor, we set the used name
+        if first_seen_name is None or params_and_buffers[n] is params_and_buffers[first_seen_name]:
+            weight_to_name_and_tied_names[t] = (n, weight_to_name_and_tied_names[t][1])
+            return
+
+        raise ValueError(f"functional_call got values for both {n} and {first_seen_name}, which are tied.")
```

Comment:
nit: should we recommend setting tied_weight=False if the user wants to explicitely untie the weights?

Reply:
Added